From 72835a6dd18e507ebb277e8687d16120371b4f6c Mon Sep 17 00:00:00 2001
From: Octavian Purdila <octavian.purdila@intel.com>
Date: Tue, 27 Dec 2011 14:12:55 +0200
Subject: vold: add wildcard for device path matching

Switching the kernel to the new sysfs layout (unselecting
CONFIG_SYSFS_DEPRECATED) breaks VolD block device recognition.

If previously uevents were generated for devpath such as

/block/mmcblk0

they are now generated using the full specific path, such as:

/devices/pci0000:0e/0000:0e:18.0/mmc_host/mmc0/mmc0:1234/block/mmcblk0

Because the full specific name contains variable IDs (in this MMC case
"1234") we can't use the full specific path entries in vold.fstab to
match the devices.

This patch adds support for prefix wildcards in device pathnames
(e.g. */block/mmcblk0).

Change-Id: I2ae94b31a52ff69343a4c80198e62489a1a49cf4

Author:    Octavian Purdila <octavian.purdila@intel.com>
Signed-off-by: Radu Moisan <radu.moisan@intel.com>
Signed-off-by: Jim Bride <jim.bride@intel.com>
---
 DirectVolume.cpp |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/DirectVolume.cpp b/DirectVolume.cpp
index 2a24376..ff41772 100644
--- a/DirectVolume.cpp
+++ b/DirectVolume.cpp
@@ -91,7 +91,7 @@ int DirectVolume::handleBlockEvent(NetlinkEvent *evt) {
 
     PathCollection::iterator  it;
     for (it = mPaths->begin(); it != mPaths->end(); ++it) {
-        if (!strncmp(dp, *it, strlen(*it))) {
+        if (((*it)[0] == '*' && strstr(dp, &(*it)[1])) || !strncmp(dp, *it, strlen(*it))) {
             /* We can handle this disk */
             int action = evt->getAction();
             const char *devtype = evt->findParam("DEVTYPE");
-- 
1.7.4.1

